File: 2020/1.js
const fs = require("fs");
const numbers = fs
  .readFileSync("./1.txt")
  .toString()
  .split("\n")
  .filter((x) => /^\d+$/.test(x))
  .map(Number);

function one(numbers, sum = 2020, skip = -1) {
  for (let i = 0; i < numbers.length; i++) {
    if (i === skip) continue;

    const number = numbers[i];
    const idx = numbers.indexOf(sum - number);

    if (idx !== -1) {
      return number * (sum - number);
    }
  }

  return null;
}

function two(numbers) {
  for (let i = 0; i < numbers.length; i++) {
    const n = one(numbers, 2020 - numbers[i], i);
    if (n) return n * numbers[i];
  }

  return null;
}

console.log(one(numbers));
console.log(two(numbers));

File: 2020/1.cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> read_file(string fname) {
  freopen(fname.c_str(), "r", stdin);
  vector<int> input;
  int l;
  while(cin >> l) {
    input.push_back(l);
  }

  return input;
}

int one(vector<int> input, int sum = 2020, int skip = -1) {
  for(int i = 0; i < input.size(); i++) {
    if(i == skip) continue;

    int n = sum - input[i];
    vector<int>::iterator it = find(input.begin(), input.end(), n);
    if(it != input.end()) {
      return input[i] * n;
    }
  }

  return 0;
}

int two(vector<int> input) {
  for(int i = 0; i < input.size(); i++) {
    int n = one(input, 2020 - input[i], i);
    if(n != 0) {
      return n * input[i];
    }
  }

  return 0;
}

int main() {
  vector<int> input = read_file("1.txt");

  cout << one(input) << endl;
  cout << two(input) << endl;
}
